### PR TITLE
(H)omotopy CGM

### PR DESCRIPTION
**Pull request by gideonite**
This is a first step towards implementing Homotopy CGMs into copt.

Based closely on the existing Frank-Wolfe implementation, this
incorporates additional linear constraints into the picture via homotopy
smoothing.

In this commit there are element-wise and row-wise constraints.

The next step is to provide randomized versions of this algorithm.

### CODE REVIEW

**Comment by gideonite** [comment]
I should add some unit tests for this as well, but I thought that now might be a good time to get feedback.

**Comment by gideonite** [comment]
Note to self: are there places where I could make the code faster? Specifically, but using numba / utils.njit? There aren't any tight loops where I am going over an array. The basic Frank-Wolfe implementation doesn't have numba stuff so maybe we can't either, but I can help thinking that there might be a way to accelerated the main for-loop of the algorithm.

File: copt/constraint.py [hunk 1]
```diff
@@ -278,14 +278,16 @@ class TraceBall:
         alpha: float
             radius of the ball.
 
+        use_eigs: use an eigenvalue solver rather than an SVD solver. This
```
Comment (fabianp) [inline copt/constraint.py:281]: Is this necessary? under the hood svds is using the same eigensolver you'r using below (well, almost, it's using the Hermitian version eigsh) (https://github.com/scipy/scipy/blob/v1.6.2/scipy/sparse/linalg/eigen/arpack/arpack.py#L1727-L1913) 

If the issue is that the tolerance is too low, I think we should use the tolerance parameter in svds instead

File: copt/loss.py [hunk 2]
```diff
@@ -296,3 +296,32 @@ def f_grad(self, x, return_gradient=True):
     def lipschitz(self):
         s = splinalg.svds(self.A, k=1, return_singular_vectors=False)[0]
         return (s * s) / self.A.shape[0] + self.alpha
+
+
+class LinearObjective:
```
Comment (fabianp) [inline copt/loss.py:301]: can you rename to LinearLoss for consistency with the rest of losses?

File: examples/homotopy/plot_homotopy.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+def full_digits():
+    mat = sio.loadmat(basepath / "data/full_clustering_mnist_digits.mat")
+    C = mat['Problem']['C'][0][0]
+    opt_val = mat['Problem']['opt_val'][0][0][0][0]
+    labels = mat['Problem']['labels'][0][0][0]
+    n_labels = len(np.unique(labels))
+    return C, n_labels, opt_val
+
+class HomotopyTrace(copt.utils.Trace):
```
Comment (fabianp) [inline examples/homotopy/plot_homotopy.py:30]: please describe (missing docstring)

File: copt/constraint.py [hunk 4]
```diff
@@ -305,22 +307,76 @@ def prox_factory(self):
...
+            ut = ut.real
+            vertex = self.alpha * np.outer(ut, ut).ravel()
+        else:
+            # use svd solver
+            ut, _, vt = splinalg.svds(u_mat, k=1)
+            vertex = self.alpha * np.outer(ut, vt).ravel()
         return vertex - x, None, None, 1.
+
+
+class RowEqualityConstraint:
```
Comment (fabianp) [inline copt/constraint.py:338]: please describe (missing docstring)

File: copt/constraint.py [hunk 5]
```diff
@@ -305,22 +307,76 @@ def prox_factory(self):
...
+        return val, grad.flatten()
+
+    def feasibility(self, x):
+        X = x.reshape(self.shape)
+        err = X.dot(self.operator) - self.offset
+        val = np.linalg.norm(err)
+        return val / self.offset_norm
+
+
+class ElementWiseInequalityConstraint:
```
Comment (fabianp) [inline copt/constraint.py:365]: please describe (missing docstring)

File: copt/constraint.py [hunk 6]
```diff
@@ -305,22 +307,76 @@ def prox_factory(self):
...
+        self.offset = offset
+        self.name = name
+        self.offset_norm = np.linalg.norm(self.offset)
+
+    def __call__(self, x):
+        X = x.reshape(self.shape)
+        z = np.matmul(X, self.operator)
+        return np.all(z == self.offset)
+
+    def smoothed_grad(self, x):
```
Comment (fabianp) [inline copt/constraint.py:351]: please describe (missing docstring)

File: copt/constraint.py [hunk 7]
```diff
@@ -305,22 +307,76 @@ def prox_factory(self):
...
+        X = x.reshape(self.shape)
+        err = X.dot(self.operator) - self.offset
+        val = np.linalg.norm(err) ** 2
+        grad = 2*np.outer(err, self.operator)
+        return val, grad.flatten()
+
+    def feasibility(self, x):
```
Comment (fabianp) [inline copt/constraint.py:358]: please describe (missing docstring)

File: copt/constraint.py [hunk 8]
```diff
@@ -305,22 +307,76 @@ def prox_factory(self):
...
+    def __init__(self, shape, offset, beta_scaling=1000, name='elementwise_inequality_constraint'):
+        self.shape = shape
+        self.offset = offset
+        self.beta_scaling = beta_scaling
+        self.name = name
+
+    def __call__(self, x):
+        return np.all(x >= self.offset)
+
+    def smoothed_grad(self, x):
```
Comment (fabianp) [inline copt/constraint.py:375]: please describe (missing docstring)

File: examples/homotopy/plot_homotopy.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+import numpy as np
+import scipy.io as sio
+from copt.constraint import (ElementWiseInequalityConstraint,
+                             RowEqualityConstraint)
+from copt.homotopy import minimize_homotopy_cgm
+from copt.loss import LinearObjective
+
+basepath = Path(".")
+def reduced_digits():
+    mat = sio.loadmat(basepath / "data/reduced_clustering_mnist_digits.mat")
```
Comment (fabianp) [inline examples/homotopy/plot_homotopy.py:15]: It would be nice if this dataset was generally available from the library (https://github.com/openopt/copt/blob/master/copt/datasets.py) and not just specific for this example. This way, we could re-use it easier from other examples

**Review by fabianp — COMMENTED** [review]
thanks for the contribution! Made a quick pass and left some early comments

File: copt/constraint.py (hunk 1 above)
Comment (gideonite) [reply copt/constraint.py:281]: Interesting. Maybe the tolerance was the issue! I found that they gave quite different solutions, with eigs reproducing what was in their paper and svds not. I was also thinking that it was strange though. Hopefully, it's the tolerance.

**Comment by gideonite** [comment]
Super, thanks you! I'll do these things and update.

**Comment by gideonite** [comment]
Rather than this clumsy `use_eigs` option that was there before, I replaced it with a new lmo, TraceSpectrahedron, which optimizes over the psd matrices with bounded trace norm. This is actually what I originally wanted but I was a bit confused.

Other than that, I added the comments, changed the names, etc. as you requested. @fabianp 

Still to do:
1. [unit tests] I don't need feedback on this (yet). I just need to do it.
2. [figure out how dataset downloading should work] I'm not sure how to proceed or whether you have an opinion. I can add the dataset to the google storage but I need access. Or, I download directly from the source. The task is clustering so I'm not sure how much of the `_load_dataset` function should be reused. @fabianp

**Comment by fabianp** [comment]
No worries about the history, it's fine if there are multiple commits

**Comment by fabianp** [comment]
another option for the dataset is to use something more standard that already includes mnist like tf-datasets, https://www.tensorflow.org/datasets/overview

**Comment by gideonite** [comment]
The only problem with that is that the dataset that we are using are the outputs of a softmax. We could train the simple model dynamically when the dataset is downloaded or download directly what was used in [MVW16]. For now, I decided to download it directly.

Ok, I think it's finally ready for merge! What do you think? @fabianp 

[MVW16] D. G. Mixon, S. Villar, and R. Ward, “Clustering subgaussian mixtures by semidefinite programming,” May 2016. <http://arxiv.org/abs/1602.06612>, GitHub Repo: <https://github.com/solevillar/kmeans_sdp>

**Comment by gideonite** [comment]
Fixed a small bug and added a unit test to test for it next time. Apparently, this unit test also improves the code coverage above the threshold for coveralls (finally).

File: copt/homotopy.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+import numpy as np
+from scipy import io as sio
+from scipy import linalg, optimize
+
+import copt
+from copt import datasets, utils
+
+EPS = np.finfo(np.float32).eps
+
+# TODO :ref:`hcgm`
```
Comment (fabianp) [inline copt/homotopy.py:18]: Please remove TODO or at least make it clearer

**Comment by fabianp** [comment]
thanks @gideonite l it all looks good to me.

@GeoffNN do you mind taking a look a this ? Thanks

**Comment by GeoffNN** [comment]
I'll look at it ASAP!

File: copt/homotopy.py (hunk 10 above)
Comment (gideonite) [reply copt/homotopy.py]: Whoops, sorry about that. Removed.

**Comment by gideonite** [comment]
Does anyone else have feedback? This has been in purgatory for a while and I'd like to get it merged. @GeoffNN @fabianp

**Review by GeoffNN — COMMENTED** [review]
A few cosmetic comments, the rest looks good to me :) Will approve when you answer @gideonite, and sorry for the wait!!

**Comment by GeoffNN** [comment]
@gideonite quick ping on this, if you want to close it up!

**Comment by gideonite** [comment]
Hi @GeoffNN, sorry for the delay. Been caught up with other things and this has fallen by the wayside. I will need some more time to get back to this but if there's still interest in getting this merged, I can probably easily address most of these issues.

**Comment by GeoffNN** [comment]
The remaining issues should be straightforward / it would also be fine to leave TODO items for a layer push :)

**Comment by gideonite** [comment]
I think it's not happening :-)
Thanks for the support

**Comment by fabianp** [comment]
LOL
